rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+    x_train, y_train, x_eval, y_eval = load_datset(arguments['input_examples'])
+
+    model = LogisticRegression()
+    model.fit(x_train, y_train, args.train_steps)
+
+    #
+    # ... perform grid search, etc, ...
+    #
+
+    # Write the trained model.
+    joblib.dump(model, argument['output_model'])
+```
+
+Given the two (containerized) applications above, our goal is to chain them into
+a 2-step pipeline which will automatically track metadata related to:
+
+*   **Artifacts**: These include the input CSV file, the output CSV file with
+    training and eval examples, and the final Model file.
+*   **Executions**: These include execution steps `MyExampleGen`, with runtime
+    execution property `num_csv_columns`, and `MyTrainer`, with runtime
+    execution property `train_steps`.
+
+Notably, since `my_example_gen.py` produces *Examples*, we make the pipeline
+understand that it is *Examples*, and record it as such in ML Metadata. Doing so
+will enable downstream components, such as Trainer, to understand that it is
+receiving *Examples*, and also realize higher-level functionality such as
+ExampleValidation. Similarly, since `my_trainer.py` produces Model, other parts
+of the pipeline should understand it to enable higher-level functionality such
+as ModelValidation and Pusher to serving.
+
+Hence, we propose to provide an extension to the DSL that allows users to
+declare inputs and outputs in terms of ML Metadata of artifacts during pipeline
+construction. The proposed DSL extension would translate the input and output
+artifacts at pipeline runtime, abstracting reading and writing artifact metadata
+from and to storage on behalf of the wrapped containerized application program.
+
+## Detailed Design
+
+We propose the following syntax for wrapping user’s containers in the DSL by
+introducing `ExecutorSpec`. This syntax complements and extends the
+[ComponentSpec](https://github.com/tensorflow/tfx/blob/4cafb7a3ddd73463d70072dade71f380e86072e6/tfx/types/component_spec.py#L73)
+design previously implemented in TFX, and generalizes
+[EXECUTOR_CLASS](https://github.com/tensorflow/tfx/blob/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/components/base/base_component.py#L66)
+attribute of components. We propose to use the same `ComponentSpec` base class
+to describe the custom container-based component’s input and output artifacts,
+and its parameters.
+
+*   Base `ExecutorSpec` class for container based component.
+
+```python
+class ExecutorSpec(with_metaclass(abc.ABCMeta, object)):
+  """Base class that specifies 'what' to execute for the component."""
+  pass
+
+
+class ExecutorClassSpec(ExecutorSpec):
+  """Execution spec for a Python class-based Executor."""
+
+  # An in-process Python Executor class, derived from TFX
+  # base_executor.Executor class.
+  executor_class = ...
+
+  def __init__(self, executor_class):
+    ...
+
+
+class ExecutorContainerSpec(ExecutorSpec):
+  """Execution specification of the container-based component."""
```

Comment:
As a Kubernetes user how am I supposed to reason about and understand ExecutorContainerSpec?

What is the underlying Kubernetes resource that it maps to? Is it a Pod? Is it a Job? Is it a Tekton Task?

As a Kubernetes user my default for batch jobs is a K8s job?
https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

That gives me a bunch of base line features for batch jobs such as backoff and retry limits. There are also TTL features that allow me to control how long jobs persist in the API Server.

Are these features present?

Reply:
This specification isn't meant to be Kubernetes specific. It is intended to capture the notion that a step in the pipeline may consist of running a container. This container may need to be executed on Kubernetes, or locally, or in some managed orchestration service in the Cloud. The pipeline user shouldn't need to worry about backoff and retry limits when authoring the component. Instead, we intend for these to be specified as part of the platform configuration. This latter part is still being designed, and isn't really being addressed in this RFC.